rj_gameplay/rj_gameplay/tactic/nmark_tactic.py

```diff
@@ -1,151 +1,90 @@
-        world_state: rc.WorldState,
-    ) -> float:
-
-        # TODO: Implement real unassigned cost function
-        return role.BIG_STUPID_NUMBER_CONST_FOR_UNASSIGNED_COST_PLS_CHANGE
-
-
-class NMarkTactic(tactic.ITactic):
-    """Marks the n closest enemies to ball with the closest robots on our team to said enemies."""
-
-    def __init__(self, n: int):
-        self.num_markers = n
-
-        # create empty mark SkillEntry for each robot
-        self.mark_list: List[tactic.SkillEntry] = [
-            tactic.SkillEntry(skills.mark.Mark()) for i in range(self.num_markers)
-        ]
-
-        # create cost func for each robot
-        self.cost_list = [marker_cost() for _ in range(self.num_markers)]
-
-    def compute_props(self):
-        pass
-
-    def create_request(self, **kwargs) -> role.RoleRequest:
-        """Creates a sane default RoleRequest.
-        :return: A list of size 1 of a sane default RoleRequest.
-        """
-        pass
-
-    def get_requests(self, world_state: rc.WorldState, props) -> Any:
-        """
-        :return: role request for n markers
-        """
-
-        if world_state is not None and world_state.ball.visible:
-            # assign n closest enemies to respective skill and role costFn
-            closest_enemies = get_closest_enemies_to_ball(self.num_markers, world_state)
-            for i in range(len(closest_enemies)):
-                self.mark_list[i].skill.target_robot = closest_enemies[i]
-                self.cost_list[i].enemy_to_mark = closest_enemies[i]
-
-        # create RoleRequest for each SkillEntry
-        role_requests = {
-            self.mark_list[i]: [
-                role.RoleRequest(role.Priority.LOW, False, self.cost_list[i])
-            ]
-            for i in range(self.num_markers)
-        }
-
-        return role_requests
-
-    def tick(
-        self, world_state: rc.WorldState, role_results: tactic.RoleResults, props: None
-    ) -> List[tactic.SkillEntry]:
-        """
-        :return: skills for the number of markers assigned from the n markers
-        """
-
-        # create list of skills based on if RoleResult exists for SkillEntry
-        skills = [
-            mark_skill_entry
-            for mark_skill_entry in self.mark_list
-            if role_results[mark_skill_entry][0]
-        ]
-
-        return skills
-
-    def is_done(self, world_state):
-        # TODO: replace all similar is_done() with a .all() and generator expr
-        # see https://www.w3schools.com/python/ref_func_all.asp
-        for mark_skill in self.mark_list:
-            if not mark_skill.skill.is_done(world_state):
-                return False
-        return True
+class NMarkTactic(stp.tactic.Tactic):
+    def __init__(self, world_state: stp.rc.WorldState, num_markers: int):
+        super().__init__(world_state)
+        self.num_markers = num_markers
+
+        self.opponents_to_mark = get_opponents_to_mark(world_state, self.num_markers)
```

Comment:
Since one marker is always going to capture the ball, is there a reason not to pass in `self.num_markers - 1` here? The last opponent returned will never be used. Then you can move the `if i == 0` logic below out of each loop